File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/binding-library/java/src/main/java/com/microsoft/azure/functions/signalr/SignalRGroupAction.java
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 */
package com.microsoft.azure.functions.signalr;

/**
 * <p>
 * SignalR group action (used with SignalR output binding)
 * </p>
 *
 * @since 1.0.0
 */
public class SignalRGroupAction {

    /**
     * Constructor
     */
    public SignalRGroupAction() {
    }

    /**
     * Constructor
     * @param action    Action to take ("add" or "remove")
     * @param groupName Group to add user to or remove user from
     * @param userId    User to add to or remove from group
     */
    public SignalRGroupAction(String action, String groupName, String userId, String connectionId) {
        this.action = action;
        this.groupName = groupName;
        this.userId = userId;
        this.connectionId = connectionId;
    }

    /**
     * Connection to add to or remove from group
     */
    public String connectionId = "";

    /**
     * User to add to or remove from group
     */
    public String userId = "";

    /**
     * Group to add user to or remove user from
     */
    public String groupName = "";

    /**
     * Action to take ("add" or "remove")
     */
    public String action = "";
}

File: sdk/signalr/Microsoft.Azure.WebJobs.Extensions.SignalRService/src/Bindings/SignalRGroupAction.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Runtime.Serialization;
using Microsoft.Azure.SignalR;
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;

namespace Microsoft.Azure.WebJobs.Extensions.SignalRService
{
    /// <summary>
    /// Class that contains parameters needed for group operations.
    /// Either the group operation on connectionId or userId is supported.
    /// If connectionId and userId are both set, it will be resolved by the following order:
    ///     1. ConnectionId
    ///     2. UserId
    /// </summary>
    [JsonObject]
    public class SignalRGroupAction
    {
        /// <summary>
        /// The id of the connection invoked in the group action.
        /// </summary>
        [JsonProperty("connectionId")]
        public string ConnectionId { get; set; }

        /// <summary>
        /// The id of the user invoked in the group action.
        /// </summary>
        [JsonProperty("userId")]
        public string UserId { get; set; }

        /// <summary>
        /// The name of the group invoked in the group action.
        /// </summary>
        [JsonProperty("groupName")]
        public string GroupName { get; set; }

        /// <summary>
        /// The group action type.
        /// </summary>
        [JsonProperty("action"), JsonRequired]
        public GroupAction Action { get; set; }

        /// <summary>
        /// The SignalR Service endpoints to which the group action is going to send. If null, all the endpoints will be sent. You can get all the available SignalR Service endpoints with <see cref="SignalREndpointsAttribute"/> input binding.
        /// </summary>
        [JsonProperty("endpoints")]
        public ServiceEndpoint[] Endpoints { get; set; }
    }

    /// <summary>
    /// The type of group action.
    /// </summary>
    [JsonConverter(typeof(StringEnumConverter))]
    public enum GroupAction
    {
        /// <summary>
        /// Add a connection or a user to a group.
        /// </summary>
        [EnumMember(Value = "add")]
        Add,

        /// <summary>
        /// Remove a connection or a user from a group.
        /// </summary>
        [EnumMember(Value = "remove")]
        Remove,

        /// <summary>
        /// Remove all the connections and users from a group.
        /// </summary>
        [EnumMember(Value = "removeAll")]
        RemoveAll
    }
}